File: homework/thinhbka/lesson-7-twopointer-heap/18. 4Sum.py
#Method1: n^3 logn
# def bSearch(ele,arr,start,end):
#     while start<=end:
#         mid = (start+end)>>1
#         mem = arr[mid]
#         if mem == ele:
#             return True
#         if mem > ele:
#             end = mid-1
#         else:
#             start = mid+1
#     return False
# class Solution:
#     def fourSum(self, nums, target):
#         n= len(nums)
#         nums.sort()
#         # print("hash table ",sum12)
#         ans = {}
#         for a in range(n):
#             for b in range(a+1,n):
#                 for c in range(b+1,n):
#                     s = target-(nums[a]+nums[b]+nums[c])
#                     if bSearch(s,nums,c+1,n-1):
#                         if (nums[a],nums[b],nums[c],s) not in ans:
#                             ans[(nums[a],nums[b],nums[c],s)] = True
#         return list(ans.keys())
#Method 2: n^3 time and O(n) space
class Solution:
    def fourSum(self, nums, target):
        n= len(nums)
        nums.sort()
        ans = set()
        def find(nums,fist,second,target):
            left = second+1
            right = len(nums)-1
            target = target - nums[fist]-nums[second] 
            while left < right:
                if nums[left]+nums[right] == target:
                    ans.add((nums[fist],nums[second],nums[left],nums[right]))
                    left+=1
                    right-=1
                elif nums[left]+nums[right]<target:
                    left+=1
                else:
                    right-=1
        for a in range(n):
            for b in range(a+1,n):
                find(nums,a,b,target)
        return list(list(x) for x in ans)
Solution().fourSum([1,0,-1,0,-2,2],
0)
Solution().fourSum([2,2,2,2,2],
8)

File: homework/Truong_Duong/Lesson_11/18_4Sum.js
// Solution for https://leetcode.com/problems/4sum/

var fourSum = function (nums, target) {
  const ans = [];
  nums.sort((x, y) => x - y);

  for (let i = 0; i < nums.length; i++) {
    const target2 = target - nums[i];

    for (let j = i + 1; j < nums.length; j++) {
      const target3 = target2 - nums[j];

      let left = j + 1;
      let right = nums.length - 1;

      while (left < right) {
        const sum = nums[left] + nums[right];
        if (sum < target3) left++;
        else if (sum > target3) right--;
        else {
          const temp = new Array(4);
          temp[0] = nums[i];
          temp[1] = nums[j];
          temp[2] = nums[left];
          temp[3] = nums[right];
          ans.push(temp);

          while (left < right && nums[left] === temp[2]) left++;
          while (left < right && nums[right] === temp[3]) right--;
        }
      }

      while (j + 1 < nums.length && nums[j + 1] === nums[j]) j++;
    }
    while (i + 1 < nums.length && nums[i + 1] === nums[i]) i++;
  }
  return ans;
};